{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "def is_interactive():\n",
    "    import __main__ as main\n",
    "    return not hasattr(main, '__file__')\n",
    "\n",
    "Param_SBMLfile = \"../data/iLiverCancer1715.xml\"\n",
    "Param_Pourcentage = 1.1\n",
    "Param_RNASeqFilename = \"../data/LiverCancerTCGA_male.csv\"\n",
    "Param_MaxIndegreeInRRG = 18\n",
    "Param_Drug = \"Sorafenib (USAN/INN);Nexavar (TN)\"\n",
    "Param_NumberOfSample = 1000\n",
    "debugon = True\n",
    "Param_DoReconstruction = False\n",
    "WithRef=False\n",
    "BIF_file=\"\"\n",
    "\n",
    "\n",
    "\n",
    "if not is_interactive():\n",
    "    import sys,getopt\n",
    "\n",
    "    optlist, args = getopt.getopt(sys.argv[1:], 'i:p:r:m:d:s:vRwP:Q:b:', [\"input\", \"percentage\",\"rna\",\"maxdegree\",\"drug\",\"samples\",\"verbose\",\"with-reconstruction\",\"with-ref\",\"probas\",\"probasref\",\"bif\"])\n",
    "\n",
    "    for o,v in optlist:\n",
    "        if  o in (\"-i\",\"--input\"):\n",
    "            Param_SBMLfile = v\n",
    "    \n",
    "        if o in (\"-r\",\"--rna\"):\n",
    "            Param_RNASeqFilename = v\n",
    "    \n",
    "        if o in (\"-p\",\"--percentage\"):\n",
    "            Param_Pourcentage = float(v)\n",
    "    \n",
    "        if o in (\"-m\",\"--maxdegree\"):\n",
    "            Param_MaxIndegreeInRRG = int(v)\n",
    "    \n",
    "        if o in (\"-d\",\"--drug\"):\n",
    "            Param_Drug = v\n",
    "        \n",
    "        if o in (\"-s\",\"--samples\"):\n",
    "            Param_NumberOfSample = int(v)\n",
    "    \n",
    "        if o in (\"-v\",\"--verbose\"):\n",
    "            debugon = True\n",
    "            \n",
    "        if o in (\"-R\",\"--with-reconstruction\"):\n",
    "            Param_DoReconstruction = True\n",
    "\n",
    "        if o in (\"-w\",\"--with-ref\"):\n",
    "            WithRef = True\n",
    "\n",
    "        if  o in (\"-P\",\"--probas\"):\n",
    "            Param_ProbaReactionsFilename = v\n",
    "            \n",
    "        if  o in (\"-Q\",\"--probasref\"):\n",
    "            Param_ProbaReactionsREFFilename = v\n",
    "\n",
    "        if o in (\"-b\",\"--bif\"):\n",
    "            BIF_file = v\n",
    "\n",
    "# internal variables\n",
    "#import shutils\n",
    "#shutils.copy(Param_SBMLfile,\"../tmp\")\n",
    "import os\n",
    "\n",
    "Param_SBMLfileBN = \"../tmp/\"+os.path.basename(Param_SBMLfile)\n",
    "\n",
    "PC_Endpoint = \"http://rdf.pathwaycommons.org/sparql\"\n",
    "RXN_EC_file = Param_SBMLfileBN.replace(\".xml\",\".rxn\")\n",
    "Param_RRGFilename = Param_SBMLfileBN.replace(\".xml\",\".rrg\")\n",
    "Param_ProbaReactionsFilename = Param_SBMLfileBN.replace(\".xml\",\".prob\")\n",
    "#whichsolver = \"gurobi\"\n",
    "whichsolver = \"cglpk\"\n",
    "\n",
    "if BIF_file==\"\":\n",
    "    BIF_file = Param_SBMLfileBN.replace(\".xml\",\".bif\")\n",
    "\n",
    "\n",
    "\n",
    "filenameCorrespondanceEnsemblGenenames = \"../tables/Correspondance_Ensembl_genename.txt\"\n",
    "filenameCorrespondanceEnsemblEnzymes = \"../tables/Correspondance_Enzymes_genenames.txt\"\n",
    "filenameCorrespondanceEnsemblTransporters = \"../tables/Correspondance_Transporters_genenames.txt\"\n",
    "filenameKEGGDrugsGenes = \"../tables/Public_KEGGDRUG_Targets.csv\"\n",
    "filenameCorrespondanceEnsemblMeanLength = \"../tables/Correspondance_Ensembl_MeanTranscriptSize.txt\"\n",
    "Param_ProbaReactionsREFFilename = Param_SBMLfileBN.replace(\".xml\",\".prob_ref\")\n",
    "\n",
    "\n",
    "# internal function can be edited\n",
    "def Binarized(x,m,s,gene=\"null\"):\n",
    "    x=log(0.0001+x)\n",
    "    InfBoundCoeff=0 #0.46 #0.46\n",
    "    SupBoundCoeff=10000000 # 0.46\n",
    "    if (x<m-InfBoundCoeff*s):\n",
    "        return 0\n",
    "    if (x>m+SupBoundCoeff*s):\n",
    "        return 1\n",
    "    return 1\n",
    "\n",
    "\n",
    "def Binarized1(x,m,s,gene=\"null\"):\n",
    "    InfBoundCoeff=0.1\n",
    "    m=AllGeneMeans.get(gene)\n",
    "    if (m):\n",
    "        if (m !=0):\n",
    "            if (x/m<InfBoundCoeff):\n",
    "                return 0\n",
    "            else:\n",
    "                return 1\n",
    "    return 0\n",
    "\n",
    "def Binarized3(x,m,s,gene=\"null\"):\n",
    "    InfBoundCoeff=1\n",
    "#    m=AllGeneMeans.get(gene)\n",
    "    m=1\n",
    "    if (m):\n",
    "        if (m !=0):\n",
    "            if (x/m<InfBoundCoeff):\n",
    "                return 0\n",
    "            else:\n",
    "                return 1\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "from timeit import default_timer as timer\n",
    "\n",
    "from pgmpy.models import BayesianModel\n",
    "from pgmpy.estimators import MaximumLikelihoodEstimator, BayesianEstimator\n",
    "from pgmpy.factors.discrete import State\n",
    "from pgmpy.sampling import BayesianModelSampling\n",
    "from pgmpy.readwrite import BIFReader, BIFWriter\n",
    "\n",
    "import sys\n",
    "\n",
    "\n",
    "import networkx as nx\n",
    "from math import log\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "# a few definition to print on stderr\n",
    "def eprint(*args, **kwargs):\n",
    "    if debugon:\n",
    "        print(*args, file=sys.stderr, **kwargs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing all the correspondance tables\n"
     ]
    }
   ],
   "source": [
    "eprint(\"Importing all the correspondance tables\")\n",
    "# Tables 1 et 2: correspondance Ensembl <-> GeneNames (à refaire en SPARQL ?)\n",
    "\n",
    "filename = filenameCorrespondanceEnsemblGenenames\n",
    "\n",
    "EnsemblToGeneNames = dict()\n",
    "GeneNamesToEnsembl = dict()\n",
    "\n",
    "EnsemblToGeneNamesList = dict()\n",
    "GeneNamesToEnsemblList = dict()\n",
    "\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[2]=t[2].replace(\"\\n\",\"\")\n",
    "    EnsemblToGeneNames[t[0]]=t[2]\n",
    "    GeneNamesToEnsembl[t[2]]=t[0]\n",
    "    if (not EnsemblToGeneNamesList.get(t[0])):\n",
    "        EnsemblToGeneNamesList[t[0]]=list()\n",
    "    if (not GeneNamesToEnsemblList.get(t[2])):\n",
    "        GeneNamesToEnsemblList[t[2]]=list()\n",
    "    EnsemblToGeneNamesList[t[0]].append(t[2])\n",
    "    GeneNamesToEnsemblList[t[2]].append(t[0])\n",
    "    \n",
    "fd.close()\n",
    "\n",
    "# Missing genename associations\n",
    "GeneNamesToEnsembl[\"TGIF\"]=GeneNamesToEnsembl[\"TGIF1\"]\n",
    "GeneNamesToEnsembl[\"CART1\"]=GeneNamesToEnsembl[\"ALX1\"]\n",
    "EnsemblToGeneNames[\"ENSG00000122718\"]=\"OR2S2\"\n",
    "EnsemblToGeneNames[\"ENSG00000155640\"] = \"C10orf12\"\n",
    "GeneNamesToEnsembl[\"C10orf12\"]=\"ENSG00000155640\"\n",
    "EnsemblToGeneNames[\"ENSG00000168260\"] = \"C14orf183\"\n",
    "GeneNamesToEnsembl[\"C14orf183\"]=\"ENSG00000168260\"\n",
    "EnsemblToGeneNames[\"ENSG00000168787\"] = \"OR12D2\"\n",
    "EnsemblToGeneNames[\"ENSG00000170803\"] = \"OR2AG1\"\n",
    "EnsemblToGeneNames[\"ENSG00000171484\"] = \"OR1B1\"\n",
    "EnsemblToGeneNames[\"ENSG00000172381\"] = \"OR6Q1\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Tables 3 et 4 : Correspondance (Enzymes,Transporters) <-> Ensembl\n",
    "\n",
    "filename = filenameCorrespondanceEnsemblEnzymes\n",
    "\n",
    "ECTCDBToEnsembl = dict()\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    L=t[1].replace(\"\\n\",\"\").split(\",\")\n",
    "    ECTCDBToEnsembl[t[0]]=list()\n",
    "    for g in L:\n",
    "        if (GeneNamesToEnsembl.get(g)):\n",
    "            ECTCDBToEnsembl[t[0]].append(GeneNamesToEnsembl.get(g))\n",
    "fd.close()\n",
    "\n",
    "\n",
    "filename = filenameCorrespondanceEnsemblTransporters\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    L=t[1].replace(\"\\n\",\"\").split(\",\")\n",
    "    ECTCDBToEnsembl[t[0]]=list()\n",
    "    for g in L:\n",
    "        if (GeneNamesToEnsembl.get(g)):\n",
    "            ECTCDBToEnsembl[t[0]].append(GeneNamesToEnsembl.get(g))\n",
    "fd.close()\n",
    "\n",
    "# Table 5 : Second filter, the genes present in the RRG must possess a RNASeq value. List of genes having a RNASeq value\n",
    "\n",
    "#Request = \"\"\"curl -o /dev/stdout \"http://bgee.org/?page=gene&gene_id=%%GENEID%%\" -q | grep \"Name\" | awk -F \"<td>|</td>|<th>|</th>\" '{for(i=1;i<NF;i++) {if ($i == \"Ensembl ID\") {printf(\"%s;;\",$(i+2));}; if ($i == \"Name\") {printf(\"%s\\\\n\",$(i+2));};          }    }'>>~/Downloads/missing.csv\"\"\"\n",
    "\n",
    "#fout=open(\"/Users/jeremiebourdon/Downloads/requete.sh\",\"w\")\n",
    "\n",
    "filename = Param_RNASeqFilename\n",
    "fd = open(filename,\"r\")\n",
    "l=fd.readline()\n",
    "RNASeqlistofgenes=list()\n",
    "#missing=\"\"\n",
    "for l in fd:\n",
    "    l=l.replace('\"','').replace('\\n','')\n",
    "    t=l.split(\",\")\n",
    "    geneid=t[0].split(\".\")[0]\n",
    "#    if not (EnsemblToGeneNames.get(geneid)):\n",
    "#        print(geneid)\n",
    "#        missing=(Request.replace(\"%%GENEID%%\",geneid))\n",
    "#        fout.write(missing+\"\\n\")\n",
    "    if (EnsemblToGeneNames.get(geneid)):\n",
    "#        RNASeqlistofgenes.append(GeneNamesToEnsembl[EnsemblToGeneNames[geneid]]) # a conversion Ensembl->Gene->Ensembl to get a unique Ensemble Id for a given gene name\n",
    "        RNASeqlistofgenes.append(geneid)\n",
    "#    else:\n",
    "#        eprint(\"Missing\",geneid)\n",
    "fd.close()\n",
    "\n",
    "#fout.close()\n",
    "\n",
    "\n",
    "# Table 6 : Drug targets\n",
    "fd=open(filenameKEGGDrugsGenes ,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "\n",
    "Allmeds=dict()\n",
    "\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[len(t)-1]=t[len(t)-1].replace(\"\\n\",\"\")\n",
    "    completeTargets=t[2].split('|')+t[3].replace(\"Enzyme:\",\"\").split(',')+t[4].split('|')\n",
    "    completeTargets=[x.strip().split(\" \")[0] for x in completeTargets if len(x)>2]\n",
    "    completeTargetsEnsembl=[GeneNamesToEnsembl.get(x) for x in completeTargets if (GeneNamesToEnsembl.get(x) != None)]\n",
    "    Allmeds[t[1].strip()]={'KeggId': t[0], 'TargetsGenes': completeTargets, 'TargetsGenesEnsembl': completeTargetsEnsembl }\n",
    "\n",
    "fd.close()\n",
    "\n",
    "# Table 6 : Mean Transcript length\n",
    "\n",
    "fd = open(filenameCorrespondanceEnsemblMeanLength,\"r\")\n",
    "\n",
    "lines=fd.readlines()\n",
    "\n",
    "AllGeneMeans = dict()\n",
    "for g in lines:\n",
    "    t=g.split(\";\")\n",
    "    t[1]=float(t[1].replace(\"\\n\",\"\"))\n",
    "    AllGeneMeans[t[0]]=t[1]\n",
    "    \n",
    "fd.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : Querying the Bayesian Network\n",
      "Execution time, first part of Querying the Bayesian Network =  0.21679582800425123 seconds\n",
      "Execution time, second part (Querying the Bayesian Network) =  354.1773161540041 seconds\n"
     ]
    }
   ],
   "source": [
    "# An alternative way is to perform a Bayesian Sampling of the BN\n",
    "\n",
    "eprint(\"Entering : Querying the Bayesian Network\")\n",
    "\n",
    "from pgmpy.readwrite import BIFReader\n",
    "model = BIFReader(BIF_file).get_model()\n",
    "\n",
    "\n",
    "def getEvidence(medname):\n",
    "    listgenes = Allmeds.get(medname).get(\"TargetsGenesEnsembl\")\n",
    "    evidences = [State(x, 0) for x in listgenes if x in model.nodes()]\n",
    "    return evidences\n",
    "\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "inference = BayesianModelSampling(model)\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time, first part of Querying the Bayesian Network = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY\n",
    "\n",
    "\n",
    "Number_of_samples = Param_NumberOfSample\n",
    "\n",
    "# NB evidences should be read somewhere\n",
    "evidences = list()\n",
    "for drug in Param_Drug.split(\";\"):\n",
    "    evidences = evidences+getEvidence(drug)\n",
    "\n",
    "# unicity of the States\n",
    "evidences = list(set(evidences))\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "samples=inference.likelihood_weighted_sample(evidences, Number_of_samples)\n",
    "#samples=inference.rejection_sample(evidences, Number_of_samples)\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time, second part (Querying the Bayesian Network) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# get all the marginal probabilities for that evidence\n",
    "allMarginalProb=dict()\n",
    "for g in model.nodes():\n",
    "    #print(g,\" p0=\",samples[g].eq(0).sum()/Number_of_samples,\" p1=\",samples[g].eq(1).sum()/Number_of_samples,\" p2=\",samples[g].eq(2).sum()/Number_of_samples)\n",
    "    allMarginalProb[g]=[samples[g].eq(0).sum()/Number_of_samples,samples[g].eq(1).sum()/Number_of_samples,samples[g].eq(2).sum()/Number_of_samples]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Correspondance Reaction <-> Enzymes\n",
    "filename = RXN_EC_file\n",
    "\n",
    "ReactionsToEnsembl = dict()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\" \")\n",
    "    RXN = t[0]\n",
    "    t[len(t)-1]=t[len(t)-1].replace(\"\\n\",\"\")\n",
    "    ReactionsToEnsembl[RXN]=list()\n",
    "    for i in range(1,len(t)):\n",
    "        val = ECTCDBToEnsembl.get(t[i].lower())\n",
    "        if (val != None):\n",
    "            ReactionsToEnsembl[RXN]+=val\n",
    "fd.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ProbaReaction(L):\n",
    "    global allMarginalProb\n",
    "    maximum=0\n",
    "    minimum=1000    \n",
    "    for l in L:\n",
    "        if allMarginalProb.get(l):\n",
    "            minimum=min([minimum,1-allMarginalProb.get(l)[0]])\n",
    "            maximum=max([maximum,1-allMarginalProb.get(l)[0]])\n",
    "    if minimum==1000:\n",
    "        return \"1;1\"\n",
    "    else:\n",
    "        return str(minimum)+\";\"+str(maximum)\n",
    "\n",
    "def ProbaReactionAll(L):\n",
    "    global allMarginalProb\n",
    "    res=list()\n",
    "    for l in L:\n",
    "        if allMarginalProb.get(l):\n",
    "#            minimum=min([minimum,1-allMarginalProb.get(l)[0]])\n",
    "#            maximum=max([maximum,1-allMarginalProb.get(l)[0]])\n",
    "            res.append(1-allMarginalProb.get(l)[0])\n",
    "        \n",
    "    return res\n",
    "\n",
    "\n",
    "\n",
    "# Output the reactions probabilities\n",
    "\n",
    "output=open(Param_ProbaReactionsFilename,\"w\")\n",
    "\n",
    "for r in ReactionsToEnsembl:\n",
    "    if (len(ReactionsToEnsembl.get(r))>0):\n",
    "        output.write(str(r)+\";\"+str(ProbaReaction(ReactionsToEnsembl.get(r)))+\"\\n\")\n",
    "output.close()\n",
    "\n",
    "# Output the reactions probabilities (all)\n",
    "\n",
    "output=open(Param_ProbaReactionsFilename+\"ALL\",\"w\")\n",
    "\n",
    "for r in ReactionsToEnsembl:\n",
    "    if (len(ReactionsToEnsembl.get(r))>0):\n",
    "        output.write(str(r)+\";\"+str(ProbaReactionAll(ReactionsToEnsembl.get(r))).replace(\",\",\";\")+\"\\n\")\n",
    "output.close()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : Graph Filtering\n"
     ]
    }
   ],
   "source": [
    "# Construction and filter\n",
    "eprint(\"Entering : Graph Filtering\")\n",
    "\n",
    "RNASeqList=dict()\n",
    "for g in RNASeqlistofgenes:\n",
    "    L=GeneNamesToEnsemblList.get(EnsemblToGeneNames[g])\n",
    "    if L:\n",
    "        for l in L:\n",
    "            RNASeqList[l]=1\n",
    "\n",
    "\n",
    "GenesToBeRemoved=list()\n",
    "\n",
    "filename = Param_RRGFilename\n",
    "\n",
    "genesingraph = dict()\n",
    "listofgenes = list()\n",
    "listofinteractions = list()\n",
    "listofinteractionsNamed = list()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "geneid=0\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[1]=t[1].replace(\"\\n\",\"\")\n",
    "#    if (GeneNamesToEnsembl[t[0]] in RNASeqlistofgenes and GeneNamesToEnsembl[t[1]] in RNASeqlistofgenes):\n",
    "    if (RNASeqList.get(GeneNamesToEnsembl[t[0]]) and RNASeqList.get(GeneNamesToEnsembl[t[1]])):\n",
    "        if (not (GeneNamesToEnsembl[t[0]] in GenesToBeRemoved or GeneNamesToEnsembl[t[1]] in GenesToBeRemoved)):\n",
    "            if (listofgenes.count(GeneNamesToEnsembl[t[0]]) == 0):\n",
    "                listofgenes.append(GeneNamesToEnsembl[t[0]])\n",
    "                genesingraph[GeneNamesToEnsembl[t[0]]]=len(listofgenes)-1\n",
    "            if (listofgenes.count(GeneNamesToEnsembl[t[1]]) == 0):\n",
    "                listofgenes.append(GeneNamesToEnsembl[t[1]])\n",
    "                genesingraph[GeneNamesToEnsembl[t[1]]]=len(listofgenes)-1\n",
    "            listofinteractions.append((genesingraph[GeneNamesToEnsembl[t[0]]],genesingraph[GeneNamesToEnsembl[t[1]]]))\n",
    "            listofinteractionsNamed.append((GeneNamesToEnsembl[t[0]],GeneNamesToEnsembl[t[1]]))\n",
    "    else:\n",
    "        eprint(\"Warning : \",GeneNamesToEnsembl[t[0]],\"or\",GeneNamesToEnsembl[t[1]],\"isn't in RNASeq\",)\n",
    "\n",
    "fd.close()\n",
    "\n",
    "\n",
    "\n",
    "# It is required to remove cycles and loops in the graph.\n",
    "# Here is a greedy method\n",
    "\n",
    "listofinteractionsNamedClean = list()\n",
    "\n",
    "RRG = nx.DiGraph()\n",
    "for t in listofinteractionsNamed:\n",
    "    if True:\n",
    "#    if not((RRG.has_node(t[0]) and RRG.has_node(t[1]) and (RRG.has_edge(t[1],t[0]) or nx.has_path(RRG,t[1],t[0])))):\n",
    "        RRG.add_edge(t[0],t[1])\n",
    "        listofinteractionsNamedClean.append((t[0],t[1]))\n",
    "\n",
    "globalmax=100\n",
    "nbRemovedGenes=0\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "906"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(RRG.nodes())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# basic optimization (use of a hash table to know if a gene is in the RRG)\n",
    "GeneDict=dict()\n",
    "\n",
    "for g in listofgenes:\n",
    "    L=GeneNamesToEnsemblList.get(EnsemblToGeneNames[g])\n",
    "    if L:\n",
    "        for l in L:\n",
    "            GeneDict[l]=g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : RNASeq data discretization\n",
      "Execution time (RNASeq data discretization) =  33.74850081499608 seconds\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# Reading the datafile (RNA-seq normalized with DESeq2)\n",
    "\n",
    "eprint(\"Entering : RNASeq data discretization\")\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "filename = Param_RNASeqFilename\n",
    "\n",
    "data = dict()\n",
    "\n",
    "TumoralConditions=list()\n",
    "NormalConditions=list()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "l=fd.readline()\n",
    "l=l.replace('\"','').replace('\\n','')\n",
    "t=l.split(\",\")\n",
    "\n",
    "for i in range(1,len(t)):\n",
    "    if (t[i][0:2] == \"CT\"):\n",
    "        TumoralConditions.append(i)\n",
    "    if (t[i][0:2] == \"CN\"):\n",
    "        NormalConditions.append(i)\n",
    "\n",
    "# à résoudre: il y a des gènes dans le graphe qui ne sont pas dans les données... Arrrgh\n",
    "\n",
    "raw_data=np.random.randint(low=0, high=1, size=(len(TumoralConditions),len(listofgenes)))\n",
    "#raw_data=np.random.randint(low=0, high=1, size=(len(NormalConditions),len(listofgenes)))\n",
    "\n",
    "sortedlistofgenes=list()\n",
    "# allgenes=list()\n",
    "\n",
    "for l in fd:\n",
    "    l=l.replace('\"','').replace('\\n','')\n",
    "    t=l.split(\",\")\n",
    "    NormalValues = [log(0.0001+float(t[x])) for x in NormalConditions]\n",
    "#    NormalValues=np.array(len(NormalConditions))\n",
    "#    for j in range(len(NormalConditions)):\n",
    "#        NormalValues[j]=t[NormalConditions[j]]\n",
    "    NormalMean=np.mean(NormalValues)\n",
    "    NormalStd=np.std(NormalValues)\n",
    "    geneid=t[0].split(\".\")[0]\n",
    "    BinarizedValues=[Binarized(float(t[x]),NormalMean,NormalStd,geneid) for x in TumoralConditions]\n",
    "    #if (GeneDict.get(geneid) == 1):\n",
    "    #        print(geneid,sum(BinarizedValues))\n",
    "    #if (geneid in listofgenes):\n",
    "    if (GeneDict.get(geneid)):\n",
    "        raw_data[:,len(sortedlistofgenes)]=BinarizedValues\n",
    "        sortedlistofgenes.append(GeneDict.get(geneid))\n",
    "\n",
    "    #print(t[0].split(\".\")[0],BinarizedValues)\n",
    "\n",
    "fd.close()\n",
    "\n",
    "data = pd.DataFrame(raw_data, columns=sortedlistofgenes)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time (RNASeq data discretization) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "L1=data.get('ENSG00000000938')\n",
    "L2=data.get('ENSG00000002726')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p=np.zeros(4)\n",
    "for i in range(0,len(L1)):\n",
    "    v=2*L1[i]+L2[i]\n",
    "    p[v]+=1\n",
    "p=[x/sum(p) for x in p]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.92063492063492069"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p[1]/(p[1]+p[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.88235294117647067"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p[0]/(p[0]+p[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  184 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000143190  at point  226 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000065534  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000065534  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  33 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  184 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  211 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125834  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  33 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  211 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  226 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000078399  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  184 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  226 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000147099  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  33 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000075891  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  211 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000103495  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000184481  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  184 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  211 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  226 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000125618  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000185551  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000185551  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000185551  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000185551  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000185551  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000095596  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000095596  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  33 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  168 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  176 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  180 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  184 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  211 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  226 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000137309  at point  238 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  16 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  26 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  33 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  84 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  87 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  114 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  125 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  127 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  146 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  166 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  214 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  217 1 2\n",
      "Error for  ENSG00000164299  and  ENSG00000087095  at point  238 1 2\n"
     ]
    }
   ],
   "source": [
    "allprobs=dict()\n",
    "allnodes = RRG.nodes()\n",
    "\n",
    "for g in allnodes:\n",
    "    Ledges=RRG.in_edges(g)\n",
    "    allprobs[g]=dict()\n",
    "    L1=data.get(g)\n",
    "    for (x,y) in Ledges:\n",
    "        L2=data.get(x)\n",
    "        p=np.zeros(4)\n",
    "        for i in range(0,len(L1)):\n",
    "            v=(2*L1[i]+L2[i])\n",
    "            if v<4:\n",
    "                p[v]+=1\n",
    "            else:\n",
    "                eprint(\"Error for \",x,\" and \",y, \" at point \",i,L1[i],L2[i])\n",
    "        p=[x/sum(p) for x in p]\n",
    "        allprobs[g][x]=p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000137273': {'ENSG00000091831': [0.068000000000000005,\n",
       "   0.016,\n",
       "   0.88800000000000001,\n",
       "   0.028000000000000001],\n",
       "  'ENSG00000095596': [0.084000000000000005,\n",
       "   0.0,\n",
       "   0.88400000000000001,\n",
       "   0.032000000000000001],\n",
       "  'ENSG00000100987': [0.016,\n",
       "   0.068000000000000005,\n",
       "   0.17199999999999999,\n",
       "   0.74399999999999999],\n",
       "  'ENSG00000118513': [0.043999999999999997,\n",
       "   0.040000000000000001,\n",
       "   0.30399999999999999,\n",
       "   0.61199999999999999],\n",
       "  'ENSG00000125952': [0.051999999999999998,\n",
       "   0.032000000000000001,\n",
       "   0.61199999999999999,\n",
       "   0.30399999999999999],\n",
       "  'ENSG00000129993': [0.084000000000000005, 0.0, 0.89200000000000002, 0.024],\n",
       "  'ENSG00000135100': [0.048000000000000001,\n",
       "   0.035999999999999997,\n",
       "   0.308,\n",
       "   0.60799999999999998],\n",
       "  'ENSG00000136997': [0.064000000000000001,\n",
       "   0.02,\n",
       "   0.73999999999999999,\n",
       "   0.17599999999999999],\n",
       "  'ENSG00000138795': [0.024, 0.059999999999999998, 0.128, 0.78800000000000003],\n",
       "  'ENSG00000159216': [0.056000000000000001,\n",
       "   0.028000000000000001,\n",
       "   0.45200000000000001,\n",
       "   0.46400000000000002],\n",
       "  'ENSG00000164093': [0.059999999999999998,\n",
       "   0.024,\n",
       "   0.46000000000000002,\n",
       "   0.45600000000000002],\n",
       "  'ENSG00000178764': [0.032000000000000001,\n",
       "   0.051999999999999998,\n",
       "   0.34000000000000002,\n",
       "   0.57599999999999996],\n",
       "  'ENSG00000257923': [0.048000000000000001,\n",
       "   0.035999999999999997,\n",
       "   0.504,\n",
       "   0.41199999999999998]}}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allprobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
